### ISSUE TITLE
Expose platform specific assemblies which are already exposed in the runtime pack in the targeting pack

### ISSUE THREAD

**Issue by ViktorHofer** [dotnet/runtime#53892]
There are five assemblies which are currently not exposed in the .NETCoreApp targeting pack because they are platform specific (Windows): https://github.com/dotnet/runtime/blob/54fdd2606170f0b0584eea78c93ae554deaacdf5/src/libraries/NetCoreAppLibrary.props#L148-L154

As the platform analyzers warns for usages of APIs which aren't available on specific platforms, it should be considered to expose the few missing reference assemblies which are already part of the runtime pack. The big benefit of doing so is that customers don't need to reference the corresponding packages to target these APIs and that their packages can be dead-ended as their APIs are de-facto frozen.

Discussed this with @ericstj and @terrajobst and they supportive. Ideally we would like to get this into Preview 6 which means we would want to reach consensus sooner than later. Thoughts on this?

cc @adamsitnik @carlossanlop @jozkee @jeffhandley for System.IO.FileSystem.AccessControl and System.IO.Pipes.AccessControl
cc @bartonjs @GrabYourPitchforks @jeffhandley for System.Security.AccessControl and System.Security.Principal.Windows
cc @Anipik and myself for Microsoft.Win32.Registry

**Comment by jkotas** [comment]
> their packages can be dead-ended as their APIs are de-facto frozen.

Why is the API of these packages de-facto frozen?

**Comment by ViktorHofer** [comment]
> Why is the API of these packages de-facto frozen?

We don't want to accidently ship new APIs in a .NETStandard or .NETFramework ref in a package that is already inbox on any platform. For .NETCoreApp, the APIs can only evolve together with the runtime as their runtime assets are inbox.

The customer benefit of exposing these platform specific assemblies in the targeting pack are:
- No need to reference a package to be able to use these APIs
- Platform analyzer support that warns when these APIs are being used on an unsupported platform instead of just a PNSE assembly throwing on runtime. Even though the APIs appear on Intellisense, the analyzer warns when they are being used.
- All of these assemblies are already exposed on upstream targeting packs: https://github.com/dotnet/runtime/blob/3b47bf83bd34409a09155483a934602c11e52c38/src/libraries/NetCoreAppLibrary.props#L156-L225. Adding these to the Microsoft.NETCore.App.Ref targeting pack as well makes targeting these APIs more consistent across our products (.NETCoreApp, AspNetCore, WindowsDesktop).

The benefits we get out doing this are:
- We can dead end the existing packages which don't change much as the APIs are already available inbox. S.S.Principal.Windows which carries along a functional .NETStandard asset will be trimmed out by conflict resolution anyway as its asset is already available inbox.
- By dead-ending the package, we don't have to support and build them anymore which lowers the complexity of the projects.
- Upstream targeting packs can remove these reference assemblies from their targeting packs and can also eliminate the package references.
- Supporting references which are exposed in packages and their runtime asset is available in the framework add complexity to the build and test infrastructure in dotnet/runtime. A recent example of that is the package testing infrastructure where the VerifyClosure task can't handle the omission of the reference assembly which would come in transitively via System.Management but as it's marked as Exclude="compile", the build doesn't see it: https://github.com/dotnet/runtime/blob/3855df9f0b7825e5da7ce144ca742309d55950a7/src/libraries/pkg/test/packageSettings/System.Management/net6.0/disableVerifyClosure.targets.

> It is benefit if you are targeting Windows. If you are targeting cross-plat, you just get more useless cruft in intelinse.

The proposal consists of adding these APIs to the Microsoft.NETCore.App.Ref targeting pack:
- https://github.com/dotnet/runtime/blob/main/src/libraries/Microsoft.Win32.Registry/ref/Microsoft.Win32.Registry.cs
- https://github.com/dotnet/runtime/blob/main/src/libraries/System.IO.FileSystem.AccessControl/ref/System.IO.FileSystem.AccessControl.cs
- https://github.com/dotnet/runtime/blob/main/src/libraries/System.IO.Pipes.AccessControl/ref/System.IO.Pipes.AccessControl.cs
- https://github.com/dotnet/runtime/blob/main/src/libraries/System.Security.AccessControl/ref/System.Security.AccessControl.cs
- https://github.com/dotnet/runtime/blob/main/src/libraries/System.Security.Principal.Windows/ref/System.Security.Principal.Windows.cs

Trying to quantify the addition, we would bring in 82 types (13+7+4+46+12) and most of those are already exposed in the aspnetcore targeting pack which is cross-plat. It would be interesting to explore if Intellisense could hide unsupported types when targeting a specific platform.

Does that make sense? If we want to go into more detail, we could meet with Immo to further discuss the addition.

**Comment by jkotas** [comment]
> We don't want to accidently ship new APIs in a .NETStandard or .NETFramework ref in a package that is already inbox on any platform. For .NETCoreApp, the APIs can only evolve together with the runtime as their runtime assets are inbox.

I do not think that this is correct statement. This restriction only applies to packages that are implemented via CoreLib forwarder inbox. It does not apply to pure packages that just happen to be included in the platform, like the packages discussed here. Other examples include System.Text.Json, System.Collections.Immutable or System.Reflection.Metadata. These packages are inbox, and we have added and we will likely add more API to the .NETStandard ref in these packages. Or is there some fundamental difference between the packages discussed here and S.T.Json, S.C.Immutable or S.R.Metadata?

I do not have a problem with doing this if we have good reasons. I have commented on this issue since the reasons stated in the top post did not look valid to me.
